File: assignments/07.08.2023/509.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<int> memo; // Memoization table

    Solution() {
        memo.resize(100, -1); // Initialize memoization table
    }

    int fibonacci(int n) {
        if (n <= 0) {
            return 0;
        } else if (n == 1) {
            return 1;
        }

        if (memo[n] != -1) {
            return memo[n]; // Return memoized result if available
        }

        memo[n] = fibonacci(n - 1) + fibonacci(n - 2); // Memoize the result
        return memo[n];
    }
};

int main() {
    Solution solution;

    int n;
    std::cout << "Enter the value of n: ";
    std::cin >> n;

    std::cout << "F(" << n << ") = " << solution.fibonacci(n) << std::endl;

    return 0;
}

File: assignments/07.08.2023/509.java
public class Solution {
    public int fib(int n) {
        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return 1;
        }
        return fib(n - 1) + fib(n - 2);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int n = 6; // Replace with the desired value of n
        int result = solution.fib(n);
        System.out.println("F(" + n + ") = " + result);
    }
}
